### PR TITLE
fix(Tianmu) Fix assertion errors caused by syntax optimization of the tianmu engine (#669)

### LINKED ISSUE #669: bug: Lost connection to MySQL server during query
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

image

### Expected behavior

_No response_

### How To Reproduce

```sql
create table t11 (a int NOT NULL, b int, primary key (a))ENGINE=TIANMU;
create table t12 (a int NOT NULL, b int, primary key (a))ENGINE=TIANMU;
create table t2 (a int NOT NULL, b int, primary key (a))ENGINE=TIANMU;
insert into t11 values (0, 10),(1, 11),(2, 12);
insert into t12 values (33, 10),(0, 11),(2, 12);
insert into t2 values (1, 21),(2, 12),(3, 23);

delete from t11 where t11.b not in (select b from t2 where t11.a < t2.a);
```
When the following  command is executed repeatedly, Lost connection to MySQL server during query

```sql
delete from t11 where t11.b not in  (select b from t2 where t11.a < t2.a);
```

### Environment

ubuntu@ubuntu:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

ubuntu@ubuntu:~$ /stonedb57/install/bin/mysqld --version
/stonedb57/install/bin/mysqld Ver 5.7.36_v1.0.1_beta_2-StoneDB for Linux on x86_64 (build-)
build information as follow:
Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
Branch name: stonedb-5.7-dev
Last commit ID: https://github.com/stoneatom/stonedb/commit/208f79d80bae1667d147e0713950317c9ca6c6b9
Last commit time: Date: Sat Sep 17 16:19:42 2022 +0800
Build time: Date: Fri Sep 23 14:14:27 CST 2022

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by konghaiya**
Fix assertion errors caused by syntax optimization of the tianmu engine (#669)

Cause of the problem:
The tianmu engine created join objects in syntax optimization, but did not release them 
Modify Scheme:
Release the join object after use

## Summary about this PR

Issue Number: close #669

## Tests Check List

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: mysql-test/suite/tianmu/t/issue669.test [hunk 1]
```diff
@@ -0,0 +1,22 @@
+--source include/have_tianmu.inc
+
+--echo #
+--echo # Fix assertion errors caused by syntax optimization of the tianmu engine #669
+--echo # Cause of the problem:
+--echo # The tianmu engine created join objects in syntax optimization, but did not explain them
+--echo # Modify Scheme:
+--echo # Release the join object after use
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue669.test]: Remove these comments, uncessary.

File: storage/tianmu/core/query_compile.cpp [hunk 2]
```diff
@@ -1043,7 +1045,11 @@ int Query::Compile(CompiledQuery *compiled_query, SELECT_LEX *selects_list, SELE
       cq->Union(prev_result, prev_result, tmp_table, union_all);
     if (sl == last_distinct) union_all = true;
     if (cond_to_reinsert && list_to_reinsert) list_to_reinsert->push_back(cond_to_reinsert);
-	sl->cleanup(0);
+    sl->cleanup(0);
+    if (isNewJon) {
+      delete sl->join;
+      sl->join = NULL;
```
Comment (RingsC) [inline storage/tianmu/core/query_compile.cpp]: using nullptr, not NULL

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/query_compile.cpp [hunk 3]
```diff
@@ -919,47 +920,44 @@ int Query::Compile(CompiledQuery *compiled_query, SELECT_LEX *selects_list, SELE
...
     // if (order) global_order = 0;   //we want to zero global order (which
     // seems to be always present) if we find a local order by clause
     //  The above is not necessary since global_order is set only in case of
     //  real UNIONs
 
-        ORDER *           group = sl->group_list.first;
-        Item *            having = sl->having_cond();
-        List<TABLE_LIST> *join_list = sl->join_list;
-        bool              zero_result = sl->join->zero_result_cause != NULL;
+    ORDER *group = sl->group_list.first;
```
Comment (RingsC) [inline storage/tianmu/core/query_compile.cpp:962]: why NULL is still here.

**Review by RingsC — CHANGES_REQUESTED** [review]

**Review by RingsC — APPROVED** [review]
LGTM

File: storage/tianmu/core/query_compile.cpp [hunk 4]
```diff
@@ -919,51 +920,48 @@ int Query::Compile(CompiledQuery *compiled_query, SELECT_LEX *selects_list, SELE
...
+    bool isNewJon = false;
+    if (!sl->join) {
+      sl->add_active_options(SELECT_NO_UNLOCK);
+      JOIN *join = new JOIN(sl->master_unit()->thd, sl);
+
+      if (!join) {
+        sl->cleanup(0);
+        return TRUE;
+      }
+      isNewJon = true;
```
Comment (hustjieke) [inline storage/tianmu/core/query_compile.cpp]: What is var `isNewJon` mean?  ---> `isNewJoin`?
Add comment to desc this var.
or use `ifNewJoinForTianmu` ?

**Review by hustjieke — APPROVED** [review]
LGTM

**Pull request merged**